crates/net/src/sim2h_worker.rs

```diff
@@ -89,120 +79,95 @@ impl Sim2hWorker {
             server_url: Url::parse(&config.sim2h_url)
                 .expect("Sim2h URL can't be parsed")
                 .into(),
             space_data: None,
             agent_id,
             conductor_api,
             time_of_last_sent: Instant::now(),
+            connection_status: ConnectionStatus::None,
         };
 
-        detach_run!(&mut instance.transport, |t| t.process(&mut instance))?;
-        detach_run!(&mut instance.transport, |t| t.process(&mut instance))?;
-        detach_run!(&mut instance.transport, |t| t.process(&mut instance))?;
-
-        let result = rx.recv()?;
-        let bound_url = result.map_err(|bind_error| err_msg(bind_error))?;
         debug!("Successfully bound to {:?}", bound_url);
 
+        instance.connection_status = instance.try_connect(Duration::from_millis(5000))?;
+
+        match instance.connection_status {
+            ConnectionStatus::Ready => info!("Connected to sim2h server!"),
+            ConnectionStatus::None => error!("Could not connect to sim2h server!"),
+            ConnectionStatus::Initializing => warn!("Still initializing connection to sim2h after 5 seconds of waitign"),
+        };
+
         Ok(instance)
     }
 
+    fn try_connect(&mut self, timeout: Duration) -> NetResult<ConnectionStatus> {
+        let url: url::Url = self.server_url.clone().into();
+        let clock = std::time::SystemTime::now();
+        let mut status: NetResult<ConnectionStatus> = Ok(ConnectionStatus::None);
+        loop {
+            match self.stream_manager.connection_status(&url) {
+                ConnectionStatus::Ready => return Ok(ConnectionStatus::Ready),
+                ConnectionStatus::None => {
+                    let url = self.server_url.clone().into();
+                    if let Err(e) = self.stream_manager.connect(&url) {
+                        status = Err(e.into());
+                    }
+                }
+                s => {
+                    status = Ok(s);
+                    let (_did_work, mut events) = self.stream_manager.process()?;
+                    self.stream_events.append(&mut events);
+                    std::thread::sleep(std::time::Duration::from_millis(10))
+                }
+            };
+            if clock.elapsed().unwrap() > timeout {
+                error!("Timed out waiting for connection for url {:?}", url);
+                return status;
+            }
+        }
+    }
+
     fn send_wire_message(&mut self, message: WireMessage) -> NetResult<()> {
         self.time_of_last_sent = Instant::now();
         let payload = wire_message_into_escaped_string(&message);
         let signature = self
             .conductor_api
-            .execute(
-                payload.clone(),
-                CryptoMethod::Sign,
-            )
-            .expect(&format!("Couldn't sign wire message in sim2h worker: {}", payload));
+            .execute(payload.clone(), CryptoMethod::Sign)
+            .expect(&format!(
```

Comment:
better to use unwrap_or_else(|| panic!("error msg here")) rather than expect(format because expect is not lazily evaluated